File: 其他功能/OC版--TbaleView索引高仿/OC版--TbaleView索引高仿/MGIndexView/View/MGIndexView.m
//
//  MGIndexView.m
//  OC版--TbaleView索引高仿
//
//  Created by i-Techsys.com on 17/4/6.
//  Copyright © 2017年 i-Techsys. All rights reserved.
//

#import "MGIndexView.h"

@class MGIndexViewDelegate;

@interface MGIndexView()
/**  存放字母按钮数组 */
@property (nonatomic,strong) NSMutableArray *letterButtons;
/**  当选中按钮 */
@property (nonatomic,strong) UIButton *selectedButton;
@end

@implementation MGIndexView

#pragma mark - Lazy
- (NSMutableArray *)letterButtons {
    if (!_letterButtons) {
        _letterButtons = [[NSMutableArray alloc] init];
    }
    return _letterButtons;
}

#pragma mark - 重写setter
- (void)setLetters:(NSMutableArray *)letters {
    _letters = letters;
    [self setUpUI];
}

- (void)setNormalTitleColor:(UIColor *)normalTitleColor {
    _normalTitleColor = normalTitleColor;
}
- (void)setSelectTitleColor:(UIColor *)selectTitleColor{
    _selectTitleColor = selectTitleColor;
}


#pragma mark - 系统方法
- (instancetype)initWithDelegate:(id<MGIndexViewDelegate>)delegate {
    if (self = [super init]) {
        self.deleagte = delegate;
        // 1.代理
        if (_deleagte != nil && [_deleagte respondsToSelector:@selector(indexViewSectionIndexTitles:)]) {
            self.letters = [NSMutableArray arrayWithArray:[_deleagte indexViewSectionIndexTitles:self]];
        }
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        self.frame = CGRectMake(MGScreenW - 20, 0, 18, MGScreenH);
        _selectedScaleAnimation = NO;
        _normalTitleColor = [UIColor grayColor];
        _selectTitleColor = [UIColor orangeColor];
        
        // 1.通知
        [MGNotificationCenter addObserver:self selector:@selector(scrollViewSelectButtonTitleColor:) name:MGWillDisplayHeaderViewNotification object:nil];
        [MGNotificationCenter addObserver:self selector:@selector(scrollViewSelectButtonTitleColor:) name:MGDidEndDisplayingHeaderViewNotification object:nil];
    }
    return self;
}

- (void)scrollViewSelectButtonTitleColor:(NSNotification *)noti {
    NSLog(@"section === %@", noti.userInfo[@"section"]);
    NSNumber *number = noti.userInfo[@"section"];
    int section = number.intValue;
    if (section >= self.letterButtons.count) { return; }
    UIButton *btn = self.letterButtons[section];

    [UIView animateWithDuration:1.0 animations:^{
        [self.selectedButton setTitleColor:self.normalTitleColor forState:UIControlStateNormal];
        self.selectedButton = btn;
        [self.selectedButton setTitleColor:self.selectTitleColor forState:UIControlStateNormal];
    }];
    
    if (self.selectedScaleAnimation) {
        btn.layer.transform = CATransform3DMakeScale(1.5, 1.5, 1.5);
        [UIView animateWithDuration:1.0 animations:^{
            btn.layer.transform = CATransform3DIdentity;
        }];
    }
}

- (void)dealloc {
    NSLog(@"MGIndexView--dealloc"); // 释放观察者
    [MGNotificationCenter removeObserver:self];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self touchesMoved:touches withEvent:event];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    UITouch *touch = touches.anyObject;
    // 获取当前的触摸点
    CGPoint curP = [touch locationInView:self];
    
    for (UIButton *btn in _letterButtons) {
        // 把btn的转化为坐标
        CGPoint btnP = [self convertPoint:curP toView:btn];
        // 判断当前点是否点在按钮上
        if ([btn pointInside:btnP withEvent:event]) {
            self.selectedButton = btn;
            [btn setTitleColor:self.selectTitleColor forState:UIControlStateNormal];
            NSInteger i = [self.letterButtons indexOfObject:btn];
            NSString *letter = btn.currentTitle;
            
            if ([_deleagte respondsToSelector:@selector(indexView:sectionForSectionIndexTitle:at:)]) {
                [_deleagte indexView:self sectionForSectionIndexTitle:letter at:i];
            }
                        
            if (_selectedScaleAnimation) { // 是否需要放大效果
                btn.layer.transform = CATransform3DMakeScale(1.5, 1.5, 1.5);
                [UIView animateWithDuration:1.0 animations:^{
                    btn.layer.transform = CATransform3DIdentity;
                }];
            }

        }else {
            [btn setTitleColor:self.normalTitleColor forState:UIControlStateNormal];
            btn.layer.transform = CATransform3DIdentity;
        }
    }
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self.selectedButton setTitleColor:self.selectTitleColor forState:UIControlStateNormal];
    if ([_deleagte respondsToSelector:@selector(indexView:cancelTouch:withEvent:)]) {
        [_deleagte indexView:self cancelTouch:touches withEvent:event];
    }
}

#pragma mark - 内部初始化方法
- (void)setUpUI {
    for (UIView *view in self.subviews) {
        [view removeFromSuperview];
    }
    [self.letterButtons removeAllObjects];
    
    
    CGFloat h = (MGScreenH == 480) ? 15 : 18;
    CGFloat x = 0;
    [self.letters enumerateObjectsUsingBlock:^(NSString*  _Nonnull letter, NSUInteger idx, BOOL * _Nonnull stop) {
        CGFloat y = idx*(h);
        UIButton *btn = [[UIButton alloc] initWithFrame:CGRectMake(x, y, self.frame.size.width, h)];
        [btn setTitle:letter forState:UIControlStateNormal];
        [btn setTitleColor:self.normalTitleColor forState:UIControlStateNormal];
        if (MGScreenH == 480) {
             btn.titleLabel.font = [UIFont systemFontOfSize:11];
        }else {
             btn.titleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleSubheadline];
        }
       
        btn.userInteractionEnabled = NO;
        [self addSubview:btn];
        [self.letterButtons addObject:btn];
        
        if (idx == 0) {
            [btn setTitleColor:self.selectTitleColor forState:UIControlStateNormal];
            self.selectedButton = btn;
        }
    }];

    [self layoutIfNeeded];
    self.height = _letterButtons.count * h;
    self.center = CGPointMake(MGScreenW-10, MGScreenH*0.5);
}

@end

File: indexView/indexView/索引条/MGIndexView/MGIndexView.swift
//
//  MGIndexView.swift
//  indexView
//
//  Created by i-Techsys.com on 17/3/13.
//  Copyright © 2017年 i-Techsys. All rights reserved.
//

import UIKit

// MARK:- 全局参数
let MGScreenBounds = UIScreen.main.bounds
let MGScreenW      = UIScreen.main.bounds.size.width
let MGScreenH      = UIScreen.main.bounds.size.height

// MARK: - 协议
@objc
protocol MGIndexViewDelegate: NSObjectProtocol {
    // 手指触摸的时候调用
    func indexView(_ indexView: MGIndexView, sectionForSectionIndexTitle title: String, at index: Int) -> Int
    // 返回数据源
    func indexViewSectionIndexTitles(for indexView: MGIndexView) -> [String]?
    // 手指离开的时候调用
    @objc optional func indexView(_ indexView: MGIndexView, cancelTouch: Set<UITouch>, with event: UIEvent?)
}


class MGIndexView: UIView {
    // MARK: - 接口属性： UIColor.orange
    var selectTitleColor: UIColor = UIColor.orange { // 选中颜色
        didSet {
           selectColor = selectTitleColor
        }
    }
    var normalTitleColor: UIColor = UIColor.gray {  // 正常颜色
        didSet {
            normalColor = normalTitleColor
        }
    }
    
    var letters: [String] = [String]() {     // 右边显示的文字数组
        didSet {
            reloadSectionIndexTitles()
        }
    }
    
    var selectedScaleAnimation: Bool = false  // 默认false，选中的标题不需要放大
    weak var delegate: MGIndexViewDelegate?   // 代理
    
    // MARK: - 私有属性
    fileprivate var selectColor: UIColor = UIColor.orange
    fileprivate lazy var normalColor: UIColor = UIColor.gray
    fileprivate lazy var letterButtons: [UIButton] = [UIButton]()
    fileprivate lazy var selectedButton: UIButton = UIButton()
    
    // MARK: - 系统方法
    convenience init(frame: CGRect?, delegate: MGIndexViewDelegate?) {
        self.init()
        self.delegate = delegate
        if delegate != nil && (delegate?.responds(to: #selector(MGIndexViewDelegate.indexViewSectionIndexTitles(for:))))! {
            self.letters = delegate!.indexViewSectionIndexTitles(for: self)!
            reloadSectionIndexTitles()
        }
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.frame = CGRect(x: MGScreenW - 20, y: 0, width: 20, height: MGScreenH)
        MGNotificationCenter.addObserver(self, selector: #selector(MGIndexView.scrollViewSelectButtonTitleColor(noti:)), name: NSNotification.Name(MGWillDisplayHeaderViewNotification), object: nil)
        MGNotificationCenter.addObserver(self, selector: #selector(MGIndexView.scrollViewSelectButtonTitleColor(noti:)), name: NSNotification.Name(MGDidEndDisplayingHeaderViewNotification), object: nil)
    }
    
    @objc fileprivate func scrollViewSelectButtonTitleColor(noti: NSNotification) {
        guard let section: Int = noti.userInfo?["section"] as? Int else { return }
        if section >= self.letterButtons.count { return }
        let btn = self.letterButtons[section]
        print(section)
        UIView.animate(withDuration: 1.5) {             
            self.selectedButton.setTitleColor(.gray, for: .normal)
            self.selectedButton = btn
            self.selectedButton.setTitleColor(.orange, for: .normal)
        }
        
        if self.selectedScaleAnimation {
            self.mg_ScaleSelectedBtn(btn: btn)
        }
    }
    
    deinit {
        print("MGIndexView == deinit")
        MGNotificationCenter.removeObserver(self)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.touchesMoved(touches, with: event)
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.first else { return }
        // 获取当前的触摸点
        let curP = touch.location(in: self)
        for btn in letterButtons {
            let btnP = self.convert(curP, to: btn) // 把btn的转化为坐标
            
            // 判断当前点是否点在按钮上
            if btn.point(inside: btnP, with: event) {
                selectedButton = btn
                btn.setTitleColor(selectColor, for: .normal)
                guard let i = letterButtons.index(of: btn) else { return }
                let letter = btn.currentTitle ?? ""
                if delegate != nil && (delegate?.responds(to: #selector(MGIndexViewDelegate.indexView(_:sectionForSectionIndexTitle:at:))))! {
                    let _ = delegate?.indexView(self, sectionForSectionIndexTitle: letter, at: i)
                }
                
                if selectedScaleAnimation {
                    mg_ScaleSelectedBtn(btn: btn)
                }
            }else {
                btn.setTitleColor(normalColor, for: .normal)
                btn.layer.transform = CATransform3DIdentity
            }
        }
    }
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesEnded(touches, with: event)
        if delegate != nil && (delegate?.responds(to: #selector(MGIndexViewDelegate.indexView(_:cancelTouch:with:))))! {
            let _ = delegate?.indexView!(self, cancelTouch: touches, with: event)
        }
    }
}

// MARK: - 设置UI
extension MGIndexView {
    fileprivate func setUpUI() {
        for view in self.subviews {
            view.removeFromSuperview()
        }
        letterButtons.removeAll()
        
        
        let h = 20
        let x = 0
        for (i,letter) in letters.enumerated() {
            let y = i*(h)
             let btn = UIButton(frame: CGRect(x: x, y: y, width: Int(self.frame.size.width), height: h))
            btn.setTitle(letter, for: .normal)
            btn.setTitleColor(normalColor, for: .normal)
            btn.titleLabel?.font = UIFont.preferredFont(forTextStyle: UIFontTextStyle.subheadline)
            btn.isUserInteractionEnabled = false
            addSubview(btn)
            letterButtons.append(btn)
            
            if i == 0 {
                btn.setTitleColor(selectColor, for: .normal)
                selectedButton = btn
            }
        }
        
        layoutIfNeeded()
        self.frame.size.height = CGFloat((letterButtons.count) * (h))
        self.center = CGPoint(x: MGScreenW-10, y: MGScreenH*0.5)
    }
}

// MARK: - 封装方法
extension MGIndexView {
    fileprivate func mg_ScaleSelectedBtn(btn: UIButton) {
        btn.layer.transform = CATransform3DMakeScale(1.5, 1.5, 1.5)
        UIView.animate(withDuration: 1.0) {
            btn.layer.transform = CATransform3DIdentity
        }
    }
    
    fileprivate func reloadSectionIndexTitles() {
        setUpUI()
    }
}


